generate-logic-legacy/src/main/kotlin/GenerateDeclarations.kt

```diff
@@ -0,0 +1,64 @@
+@file:Suppress("UnstableApiUsage")
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.Directory
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.file.FileSystemOperations
+import org.gradle.api.file.ProjectLayout
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.TaskAction
+import org.gradle.kotlin.dsl.listProperty
+import org.gradle.kotlin.dsl.property
+import java.io.File
+import javax.inject.Inject
+
+abstract class GenerateDeclarationsTask
+@Inject constructor(
+    private val projectLayout: ProjectLayout,
+    private val fileSystemOperations: FileSystemOperations,
+    objectFactory: ObjectFactory,
+) : DefaultTask(), HasGeneratedDirectories {
+
+    private val rootBuildDirectory: Directory
+        get() = project.rootProject.layout.buildDirectory.get()
```

Comment:
In my understanding, using of Properties is more preferable because Gradle supports different mechanics that required lazy evaluation. E.g. if you want to customize some task properties, your customization can have no effect because you already read the non-property task parameter in the task constructor.